posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,196 @@
+
+The dispatch using ``__array_namespace__`` method is:
+* opt-in for users,  
+The new dispatching works only when users choose to use the new Array API compatible modules.
+* explicit,  
+The dispatched functions are determined via the method of the array object user creates.
+* with local control.  
+Which implementation to use is determined by calling methods on the direct arguments of a function.
+
+And the `uarray` dispatcher is:
+* opt-out for users,
+* can be implemented to be explicit or implicit,
+* can be implemented to have local, non-local or global control.
+
+
+
+<p align="center">
+    <img
+     alt="A diagram outlining CuPy support for SciPy and scikits."
+     src="/images/2021/11/CuPy_support_scipy_scikits_with_details.png">
+    <i>
+    Proposed dispatch mechanism layers for enabling CuPy and Dask support.
+    This will also support any other array library with the same Array API standard support and `uarray` backends.
+    </i>
+</p>
+
+## Array API dispatching system
+
+Array API is a standard specification of functionality that array libraries need to implement. There is a [dedicated test suite](https://github.com/data-apis/array-api-tests) to verify the compliance of specific implementations with the standard, so it's easy to check all libraries adopting the standard to have a unified behavior. NEP 47 has a couple of examples on [usage of the Array API namespace](https://numpy.org/neps/nep-0047-array-api-standard.html#adoption-in-downstream-libraries).
+
+`__array_namespace__` is a central piece of the Array API dispatch method. It is a method of an array object that returns an object that has all the array API functions on it. In the case of NumPy calling `__array_namespace__() ` would return `numpy.array_api` module. The `__array_namespace__` method attached to the array object enables downstream libraries to consume multiple kinds of arrays, without having to have a hard dependency on any of those array libraries.
+The pattern to support multiple array libraries is intended to be similar to the following code snippet:
+```python
+def somefunc(x, y):
+    # Retrieves standard namespace.
+    # Raises if x and y have different namespaces.
+    xp = get_namespace(x, y)
+    out = xp.mean(x, axis=0) + 2*xp.std(y, axis=0)
+    return out
+```
+
+Notice that to make the above code valid, we don't need to import the package that implements mean and std functions explicitly; the required module is stored in x and y variables.
+
+For several concrete use cases of the Array API check out [the dedicated page](https://data-apis.org/array-api/latest/use_cases.html) of the standard. Anirudh Dagar also [wrote a great demo](https://anirudhdagar.ml/array-api-demo/GW_Demo_Array_API.html) recreating gravitational waves tutorial and showcasing what it would take to support PyTorch tensors for a subset of SciPy API using `__array_namespace__` .
```

Comment:
```suggestion
For several concrete use cases of the Array API check out [the dedicated page](https://data-apis.org/array-api/latest/use_cases.html) of the standard. Anirudh Dagar also [wrote a great demo](https://quansight-labs.github.io/array-api-demo/GW_Demo_Array_API.html) recreating gravitational waves tutorial and showcasing what it would take to support PyTorch tensors for a subset of SciPy API using `__array_namespace__` .
```
Fix link to demo.